internal/controllers/configuration/konglicense_controller.go

```diff
@@ -0,0 +1,368 @@
+	err := r.Client.Get(ctx, k8stypes.NamespacedName{Name: license.Name}, license)
+	if err != nil {
+		return fmt.Errorf("failed to get latest version of KongLicense %s: %w", license.Name, err)
+	}
+
+	fullControllerName := LicenseControllerType + ":" + r.ControllerName
+	// Find the managed controller status item and append new item when absent.
+	controllerStatus, controllerIndex, found := lo.FindIndexOf(license.Status.KongLicenseControllerStatuses, func(controllerStatus kongv1alpha1.KongLicenseControllerStatus) bool {
+		return controllerStatus.ControllerName == fullControllerName
+	})
+	if !found {
+		wantedControllerStatus := kongv1alpha1.KongLicenseControllerStatus{
+			ControllerName: fullControllerName,
+			Conditions: []metav1.Condition{
+				{
+					Type:               ConditionTypeProgrammed,
+					LastTransitionTime: metav1.Now(),
+					Status:             programmedStatus,
+					Reason:             reason,
+				},
+			},
+		}
+		license.Status.KongLicenseControllerStatuses = append(license.Status.KongLicenseControllerStatuses, wantedControllerStatus)
+		return r.Client.Status().Update(ctx, license)
+	}
+
+	wantedCondition := metav1.Condition{
+		Type:               ConditionTypeProgrammed,
+		LastTransitionTime: metav1.Now(),
+		Status:             programmedStatus,
+		Reason:             reason,
+		Message:            message,
+	}
+	// Find the "programmed" condition in the condition list.
+	programmedCondition, conditionIndex, found := lo.FindIndexOf(controllerStatus.Conditions, func(condition metav1.Condition) bool {
+		return condition.Type == ConditionTypeProgrammed
+	})
+	if !found {
+		// Return error if the number of conditions already reached the limit.
+		if len(license.Status.KongLicenseControllerStatuses) >= maxConditionNum {
+			return fmt.Errorf("already %d condition items in controller status exceeding the limit %d, cannot add new item",
+				len(license.Status.KongLicenseControllerStatuses), maxConditionNum)
+		}
+		license.Status.KongLicenseControllerStatuses[controllerIndex].Conditions = append(
+			license.Status.KongLicenseControllerStatuses[controllerIndex].Conditions, wantedCondition,
+		)
+		return r.Client.Status().Update(ctx, license)
+	}
+	if programmedCondition.Status != programmedStatus {
+		license.Status.KongLicenseControllerStatuses[controllerIndex].Conditions[conditionIndex] = wantedCondition
+		return r.Client.Status().Update(ctx, license)
+	}
+
+	return nil
+}
+
+// GetLicense is the interface to get the license in Kong configuration format to use in translator.
+func (r *KongV1Alpha1KongLicenseReconciler) GetLicense() mo.Option[kong.License] {
+	chosenLicense := r.getChosenLicense()
+	if chosenLicense == nil {
+		r.Log.V(util.DebugLevel).Info("No KongLicense available")
+		return mo.None[kong.License]()
+	}
+	r.Log.V(util.DebugLevel).Info("Get license from KongLicense resource", "name", chosenLicense.Name)
+	// TODO: Validate KongLicense on Kong gateway.
```

Comment:
This should rather happen in the webhook, no?

Reply:
Yes, I will implement in a webhook later (may not be in 3.1?)